optuna/samplers/_tpe/sampler.py

```diff
@@ -635,13 +694,44 @@ def _get_observation_pairs(
                 param_value = None
             values[param_name].append(param_value)
 
-    return values, scores
+        if get_constraints:
+            assert constraints is not None
+            constraints.append(cast(Sequence[float], trial.system_attrs.get(_CONSTRAINTS_KEY)))
+
+    return values, scores, constraints
 
 
 def _split_observation_pairs(
     loss_vals: List[Tuple[float, List[float]]],
     n_below: int,
+    constraints: Optional[List[Sequence[float]]],
 ) -> Tuple[np.ndarray, np.ndarray]:
+    # When constrains is not None, trials are split into below and above
+    # according to the following rules.
+    # 1. Feasible trials are better than infeasible trials.
+    # 2. Infeasible trials are sorted by sum of how much they violate each constraint.
+    # 3. Feasible trials are sorted by loss_vals.
+    if constraints is not None:
+        # 1-dimensional violation value (violation_1d==0 <=> feasible, >0 <=> infeasible).
+        violation_1d = np.maximum(np.array(constraints), 0).sum(1)
```

Comment:
If `constraints` includes `None`, the type of each item in `np.array(constraints)` includes `None`. How about specifying the dtype like `np.array(constraints, dtype=float)`? It convert `None` to `np.nan`, so we need to filter `np.nan` before taking `np.maximum` to compare `violation_1d[idx[n_below]] > 0`.

Reply:
I did not consider the case `constraints` includes `None`. When `constraints` include `None`, it cannot be simply converted using `np.array` because `constraints` might be something like `[(1.0, 1.0), None, (2.0, 2.0)]` and the dimension does not match. I need to consider how to handle this.

Reply:
I decided to handle `None` inside `_get_observation_pairs`. This makes `_get_observation_pairs` a bit more complex, but the whole logic and return type are simplified if we need to consider `None`. 
https://github.com/optuna/optuna/pull/3506/commits/2f1ff5a57ebe8f09013534cfe311978afc9ab14c#diff-2cc3a4088f95a10e806b68655d375f6f7f5fe2de161a0182c16168e378be8065R713-R715